feat: add diagram plugin blog post (#102)
hamlet-io/docs: blog/2021-02-26-diagram-plugin.md

---
title: Solution Diagram Generation
---
import Admonition from 'react-admonitions';

> Instead of defining the raw infrastructure you need hamlet works off a solution. The solution outlines the functional components you need in your deployment and hamlet then takes this to build out the infrastructure definition that would perform this function.

Reviewer: ```suggestion
Instead of defining the raw infrastructure you need, hamlet works off a solution. The solution outlines the functional components you need in your deployment and hamlet then takes this to build out the infrastructure definition that would perform this function.
```

Combining this with our pluggable architecture we can use the same solution to generate infrastructure definitions in different cloud providers like AWS and Azure. This is an interesting feature itself but in this post we are going to look at another way to use this data, diagram generation.

## A picture paints a 1000 words

> Declarative infrastructure deployments, which outline the exact configuration you require for your infrastructure, have become the standard with hyperscale cloud providers, they can be pretty hard to conceptually understand what you have deployed.

Reviewer: ```suggestion
Declarative infrastructure deployments that outline the precise configuration you require for your infrastructure, have become the standard with hyperscale cloud providers. However they can also be quite challenging to conceptualise what you have deployed.
```

Once the infrastructure has been deployed a common practice is to create diagrams which represent what you've deployed. These are mostly generated manually or you can use tools like cloudcraft which scan your deployment to build out the diagram. This can be a useful approach as a validation tool, however they can sometimes miss out on contextual information or even infrastructure which isn't supported by the tool.

> With hamlet our diagrams are generated using the same solution file that defined the infrastructure. This means we keep the information in lockstep between whats deployed and what has been documented, and we can also include solution level information like environments, standard naming and mutli provider support

Reviewer: ```suggestion
With hamlet our diagrams are generated using the same solution file that defined the infrastructure. This means we keep the information in lockstep between whats deployed and what has been documented, and we can also include solution level information like environments, standard naming and multi-provider support
```

## Get to it

So lets see what this thing can do

1. Start up a hamlet container and mount the container home drive to a local directory

    ```bash
    docker run -it -v ~/hamlet_diagrams/:/home/hamlet/cmdb hamletio/hamlet
    ```

    :::info
    The `~/hamlet_diagrams/` can be whatever works for you on your local system.
    We need the local mount so that we can view the diagram files created ( they will be png files )
    :::

2. Lets have a look at the diagram types that are available from hamlet. A diagram type is the structure of the diagram that you want to generate. Once you have the type you can then define a diagram with the solution contents you want,

    ```bash

>     hamlet visual -i mock -p diagramstest -p diagrams list-diagram-types

Reviewer: Perhaps the table and the command should be in separate code blocks? Or a line between them?

To use the "copy" button provided by the code block, I've now got some stuff to remove.

    ╒════╤═══════════╤══════════════════════════════════════════════════════════════════════════════════════════╕
    │    │ Type      │ Description                                                                              │
    ╞════╪═══════════╪══════════════════════════════════════════════════════════════════════════════════════════╡
    │  0 │ solution  │ A solution level diagram focussing on components and their relationships                 │
    ├────┼───────────┼──────────────────────────────────────────────────────────────────────────────────────────┤
    │  1 │ resources │ Shows all resources that are used in a segment and the deployment group they are used in │
    ╘════╧═══════════╧══════════════════════════════════════════════════════════════════════════════════════════╛
    ```

    :::info
    This post is going to use the diagramstest provider to define our solution. If you have your own CMDB you can follow along but remove the `-i mock -p diagramstest` arguments from any hamlet commands
    :::

>     The diagrams provider currently has two diagrams type, the solution which shows you all your components and their relationships based on their links, and resources which shows you the different kind of infrastructure resources in used in your solution

Reviewer: ```suggestion
    The diagrams provider currently has two diagrams type, the solution which shows you all your components and their relationships based on their links, and resources which shows you the different infrastructure resource types contained in your solution
```

3. Now lets have a look at what diagrams are available in this solution

    ```bash
    hamlet visual -i mock -p diagramstest -o diagrams list-diagrams

    ╒════╤════════════════╤═══════════╤═════════════════════════════════════════════════════════════════╕
    │    │ Id             │ Type      │ Description                                                     │
    ╞════╪════════════════╪═══════════╪═════════════════════════════════════════════════════════════════╡
    │  0 │ overview       │ solution  │ A solution level diagram of all components in the solution      │
    ├────┼────────────────┼───────────┼─────────────────────────────────────────────────────────────────┤
    │  1 │ resources_used │ resources │ Provides a diagram of all unique resources used in the solution │
    ╘════╧════════════════╧═══════════╧═════════════════════════════════════════════════════════════════╛
    ```

>     We have two digrams that we can draw in the test provider, overview, which uses the solution diagram type to show all the components, and resources_used which shows us all the resources in our solution.

Reviewer: ```suggestion
    We have two digrams that we can draw in the test provider, overview, which uses the solution diagram type to show all the **components**, and **resources_used** which shows us all the resources in our solution.
```

>     :::info

Reviewer: The article has started off from the tone of assuming the reader is following along. I think we should either continue that assumption here and instruct what to add as a part of the main article, or use an "info" admonition to link off to the module's Github page where the README can do its thing.

Author: Moved the doing it in your own solution to a section at the bottom

    Add the following to your CMDB to install the diagrams plugin and enabled the module

    ```json
    {
        "Product" : {
            "Plugins" : {
                "diagrams" : {
                    "Enabled" : true,
                    "Name" : "diagrams",
                    "Priority" : 100,
                    "Required" : true,
                    "Source" : "git",
                    "Source:git" : {
                        "Url" : "https://github.com/hamlet-io/engine-plugin-diagrams",
                        "Ref" : "main",
                        "Path" : "diagrams/"
                    }
                }
            }
        },
        "Solution" : {
            "Modules" : {
                "diagrams_overviews" : {
                    "Provider" : "diagrams",
                    "Name" : "overviews"
                }
            }
        }
    }
    ```

    Then run `${GENERATION_DIR}/setup.sh` to install the plugin
    :::

4. Now lets draw these diagrams to see what we get

    ```bash
    hamlet visual -i mock -p diagramstest -p diagrams draw-diagrams -d /home/hamlet/cmdb
    [*] overview
    [*] resources_used
    ```

>     In the local directory you mounted to the docker container you should now see two png files. One called `diagram-overview.png` and one called `diagram-resources_used.png`

Reviewer: To keep the tone consistent throughout I recommend using "We" and "Our" as much as possible over "You". 

For similar reasons I'd remove "should" and always assume all has gone well. The user can back-track on their own if they aren't seeing what you're describing.

Author: Actually ended up removing most of these, they were just extra fluff words in most cases

    ![Overview diagram showing all of the components in your solution and their links](2021-02-26-diagram-plugin/diagram-overview.png "Overview diagram")

    This diagram outlines the different components that we have, each group is a component with each of their occurrences drawn inside them.

>     - The `application-apphost-ecs` shows us an container host, *apphost* with a single container service called *webapp*
>     - The *https* load balancer rule in the *elb-webapp-lb* component has a link to this container
>     - The *https* load balancer as an incoming link from the *internet*.

Reviewer: I intended to comment here to say that the italics doesn't draw as much attention in the site as you'd think (I'm just realising this now) so recommend we bold them instead, but I suspect that was the intention all along. It appears Github will italicise the single asterisk

```suggestion
    - The `application-apphost-ecs` shows us an container host, **apphost** with a single container service called **webapp**
    - The **https** load balancer rule in the **elb-webapp-lb** component has a link to this container
    - The **https** load balancer as an incoming link from the *internet*.
```

    ![Resource types used as part of the solution](2021-02-26-diagram-plugin/diagram-resources_used.png "Resources Used")

    This diagrams shows the different infrastructure resources that we use in the solution overall.
    - At the application level we have function, container and apigateway
    - At the solution level we have a server, loadbalancer, load balancer rule and a database instance

## Where did that come from?

> So that was pretty simple, we got some diagrams they show us whats deployed and they are pretty easy to follow.

Reviewer: ```suggestion
So that was pretty simple! We made some diagrams, they show us what's deployed and they are pretty easy to follow.
```
